File: app/src/main/java/de/xikolo/services/DownloadService.java
package de.xikolo.services;

import android.app.Notification;
import android.app.Service;
import android.content.Intent;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.os.Process;
import android.util.Log;

import com.coolerfall.download.DownloadCallback;
import com.coolerfall.download.DownloadManager;
import com.coolerfall.download.DownloadRequest;
import com.coolerfall.download.OkHttpDownloader;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import de.xikolo.config.Config;
import de.xikolo.events.DownloadCompletedEvent;
import de.xikolo.models.Download;
import de.xikolo.storages.ApplicationPreferences;
import de.xikolo.utils.NotificationUtil;
import okhttp3.OkHttpClient;

public class DownloadService extends Service {

    public static final String TAG = DownloadService.class.getSimpleName();

    public static final String ARG_TITLE = "title";

    public static final String ARG_URL = "url";

    public static final String ARG_FILE_PATH = "file_path";

    private ServiceHandler serviceHandler;

    private NotificationUtil notificationUtil;

    private DownloadManager downloadClient;

    private ConcurrentHashMap<Integer, Download> downloadMap;

    private static DownloadService instance;

    public static DownloadService getInstance() {
        return instance;
    }

    private AtomicInteger jobCounter;

    @Override
    public void onCreate() {
        instance = this;

        if (Config.DEBUG) Log.d(TAG, "DownloadService created");

        jobCounter = new AtomicInteger();

        notificationUtil = new NotificationUtil(this);

        // Don't use HttpLoggingInterceptor, crashes with OutOfMemoryException!
        OkHttpClient client = new OkHttpClient.Builder()
                .build();
        downloadClient = new DownloadManager.Builder()
                .context(this)
                .downloader(OkHttpDownloader.create(client))
                .build();

        downloadMap = new ConcurrentHashMap<>();

        // Start up the thread running the service. Note that we create a
        // separate thread because the service normally runs in the process's
        // main thread, which we don't want to block.  We also make it
        // background priority so CPU-intensive work will not disrupt our UI.
        HandlerThread thread = new HandlerThread("ServiceStartArguments", Process.THREAD_PRIORITY_BACKGROUND);
        thread.start();

        // Get the HandlerThread's Looper and use it for our Handler
        Looper serviceLooper = thread.getLooper();
        serviceHandler = new ServiceHandler(serviceLooper);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (Config.DEBUG) Log.d(TAG, "DownloadService start command received");

        if (intent != null && intent.getExtras() != null && intent.getExtras().getString(ARG_TITLE) != null) {
            List<String> titles = getRunningDownloadTitles();
            titles.add(intent.getExtras().getString(ARG_TITLE));

            Notification notification = notificationUtil.getDownloadRunningNotification(titles).build();
            startForeground(NotificationUtil.DOWNLOAD_RUNNING_NOTIFICATION_ID, notification);

            // For each start request, send a message to start a job and deliver the
            // start ID so we know which request we're stopping when we finish the job
            Message msg = serviceHandler.obtainMessage();
            msg.setData(intent.getExtras());
            jobCounter.incrementAndGet();
            serviceHandler.sendMessage(msg);
        }

        // If we get killed, after returning from here, restart
        return START_STICKY;
    }

    private synchronized void stopJob() {
        if (jobCounter.decrementAndGet() == 0) {
            stopSelf();
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        // We don't provide binding, so return null
        return null;
    }

    @Override
    public void onDestroy() {
        instance = null;

        if (Config.DEBUG) Log.d(TAG, "DownloadService destroyed");

        stopForeground(true);
        downloadClient.cancelAll();
        notificationUtil = null;
    }

    public synchronized boolean isDownloading(String url) {
        Download download = getDownload(url);
        return download != null && (download.state == Download.State.PENDING || download.state == Download.State.RUNNING);
    }

    public synchronized Download getDownload(String url) {
        if (downloadMap == null || url == null) return null;

        for (Download download : downloadMap.values()) {
            if (url.equals(download.url)) return download;
        }

        return null;
    }

    public synchronized void cancelDownload(String url) {
        if (Config.DEBUG) Log.i(TAG, "Download cancelled: " + url);

        Download download = getDownload(url);
        if (download != null && isDownloading(url)) {
            downloadClient.cancel(download.id);
            downloadMap.remove(download.id);
            stopJob();
        }
    }

    private synchronized List<String> getRunningDownloadTitles() {
        List<String> titles = new ArrayList<>();

        for (Download download : downloadMap.values()) {
            if (isDownloading(download.url)) {
                titles.add(download.title);
            }
        }

        return titles;
    }

    private synchronized void updateDownloadProgress(int downloadId, long bytesWritten, long totalBytes) {
        Download download = downloadMap.get(downloadId);

        if (download != null && downloadMap != null) {
            download.state = Download.State.RUNNING;
            download.bytesWritten = bytesWritten;
            download.totalBytes = totalBytes;

            downloadMap.put(download.id, download);
        }

        long bytesWrittenOfAll = 0;
        long totalBytesOfAll = 0;

        for (Download d : downloadMap.values()) {
            bytesWrittenOfAll += d.bytesWritten;
            totalBytesOfAll += d.totalBytes;
        }

        // Race condition NullPointerException can be thrown
        if (notificationUtil != null) {
            Notification notification = notificationUtil.getDownloadRunningNotification(getRunningDownloadTitles())
                    .setProgress(100, (int) (bytesWrittenOfAll / (totalBytesOfAll / 100.)), false)
                    .build();
            notificationUtil.notify(NotificationUtil.DOWNLOAD_RUNNING_NOTIFICATION_ID, notification);
        }
    }

    private synchronized void updateDownloadSuccess(int downloadId) {
        Download download = downloadMap.get(downloadId);

        if (download != null) {
            download.state = Download.State.SUCCESSFUL;
            notificationUtil.showDownloadCompletedNotification(download);
            EventBus.getDefault().post(new DownloadCompletedEvent(download.url));
        }
    }

    private synchronized void updateDownloadFailure(int downloadId) {
        Download download = downloadMap.get(downloadId);

        if (download != null) {
            download.state = Download.State.FAILURE;
        }
    }

    // Handler that receives messages from the thread
    private final class ServiceHandler extends Handler {

        ServiceHandler(Looper looper) {
            super(looper);
        }

        @Override
        public void handleMessage(Message message) {
            final String title = message.getData().getString(ARG_TITLE);
            final String url = message.getData().getString(ARG_URL);
            final String filePath = message.getData().getString(ARG_FILE_PATH);

            int allowedNetworkTypes = DownloadRequest.NETWORK_WIFI | DownloadRequest.NETWORK_MOBILE;
            ApplicationPreferences appPreferences = new ApplicationPreferences();
            if (appPreferences.isDownloadNetworkLimitedOnMobile()) {
                allowedNetworkTypes = DownloadRequest.NETWORK_WIFI;
            }

            DownloadRequest request = new DownloadRequest.Builder()
                    .url(url)
                    .retryTime(1)
                    .progressInterval(1, TimeUnit.SECONDS)
                    .allowedNetworkTypes(allowedNetworkTypes)
                    .destinationFilePath(filePath)
                    .downloadCallback(new DownloadCallback() {
                        @Override
                        public void onStart(int downloadId, long totalBytes) {
                            if (Config.DEBUG) Log.i(TAG, "Download started: " + url);

                            updateDownloadProgress(downloadId, 0, totalBytes);
                        }

                        @Override
                        public void onProgress(int downloadId, long bytesWritten, long totalBytes) {
                            updateDownloadProgress(downloadId, bytesWritten, totalBytes);
                        }

                        @Override
                        public void onSuccess(int downloadId, String filePath) {
                            if (Config.DEBUG) Log.i(TAG, "Download finished: " + filePath);

                            updateDownloadSuccess(downloadId);

                            stopJob();
                        }

                        @Override
                        public void onFailure(int downloadId, int statusCode, String errMsg) {
                            Log.e(TAG, "Download failed: " + errMsg);

                            updateDownloadFailure(downloadId);

                            stopJob();
                        }
                    })
                    .build();

            Download download = new Download();
            download.title = title;
            download.url = url;
            download.filePath = filePath;
            download.id = downloadClient.add(request);

            downloadMap.putIfAbsent(download.id, download);
        }

    }

}

File: app/src/main/java/de/xikolo/services/DownloadService.kt
package de.xikolo.services

import android.app.Service
import android.content.Intent
import android.os.*
import android.util.Log
import com.coolerfall.download.DownloadCallback
import com.coolerfall.download.DownloadManager
import com.coolerfall.download.DownloadRequest
import com.coolerfall.download.OkHttpDownloader
import de.xikolo.App
import de.xikolo.config.Config
import de.xikolo.models.Download
import de.xikolo.network.ApiService
import de.xikolo.storages.ApplicationPreferences
import de.xikolo.utils.NotificationUtil
import okhttp3.OkHttpClient
import java.io.File
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

class DownloadService : Service() {

    companion object {
        val TAG: String = DownloadService::class.java.simpleName

        const val ARG_TITLE = "title"

        const val ARG_URL = "url"

        const val ARG_FILE_PATH = "file_path"

        const val ARG_SHOW_NOTIFICATION = "show_notification"

        var instance: DownloadService? = null
            private set
    }

    private var serviceHandler: ServiceHandler? = null

    private var notificationUtil: NotificationUtil? = null

    private var downloadClient: DownloadManager? = null

    private var downloadMap: ConcurrentHashMap<Int, Download>? = null

    private var jobCounter: AtomicInteger? = null

    val isDownloading: Boolean
        @Synchronized get() = downloadMap?.isNotEmpty() == true

    private val runningDownloadTitles: MutableList<String>
        @Synchronized get() {
            val titles = ArrayList<String>()

            downloadMap?.values?.let { downloads ->
                for (download in downloads) {
                    if (isDownloading(download.url) && download.showNotification) {
                        titles.add(download.title)
                    }
                }
            }

            return titles
        }

    override fun onCreate() {
        instance = this

        if (Config.DEBUG) Log.d(TAG, "DownloadService created")

        jobCounter = AtomicInteger()

        notificationUtil = NotificationUtil(this)

        // Don't use HttpLoggingInterceptor, crashes with OutOfMemoryException!
        val client = OkHttpClient.Builder()
            .addInterceptor(ApiService.authenticationInterceptor)
            .addInterceptor(ApiService.userAgentInterceptor)
            .build()
        downloadClient = DownloadManager.Builder()
            .context(this)
            .downloader(OkHttpDownloader.create(client))
            .build()

        downloadMap = ConcurrentHashMap()

        // Start up the thread running the service. Note that we create a
        // separate thread because the service normally runs in the process's
        // main thread, which we don't want to block.  We also make it
        // background priority so CPU-intensive work will not disrupt our UI.
        val thread = HandlerThread("ServiceStartArguments", Process.THREAD_PRIORITY_BACKGROUND)
        thread.start()

        // Get the HandlerThread's Looper and use it for our Handler
        val serviceLooper = thread.looper
        serviceHandler = ServiceHandler(serviceLooper)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (Config.DEBUG) Log.d(TAG, "DownloadService start command received")

        intent?.extras?.getString(ARG_TITLE)?.let { title ->
            if (intent.extras?.getBoolean(ARG_SHOW_NOTIFICATION, true) == true) {
                val titles = runningDownloadTitles
                titles.add(title)

                val notification = notificationUtil?.getDownloadRunningNotification(titles)?.build()
                startForeground(NotificationUtil.DOWNLOAD_RUNNING_NOTIFICATION_ID, notification)
            }

            // For each start request, send a message to start a job and deliver the
            // start ID so we know which request we're stopping when we finish the job
            serviceHandler?.obtainMessage()?.let { msg ->
                msg.data = intent.extras
                jobCounter?.incrementAndGet()
                serviceHandler?.sendMessage(msg)
            }
        }

        // If we get killed, after returning from here, restart
        return Service.START_STICKY
    }

    @Synchronized
    private fun stopJob() {
        if (jobCounter?.decrementAndGet() == 0) {
            stopSelf()
        }
    }

    override fun onBind(intent: Intent): IBinder? {
        // We don't provide binding, so return null
        return null
    }

    override fun onDestroy() {
        instance = null

        if (Config.DEBUG) Log.d(TAG, "DownloadService destroyed")

        stopForeground(true)
        downloadClient?.cancelAll()
        notificationUtil = null
    }

    @Synchronized
    fun isDownloading(url: String?): Boolean {
        val download = getDownload(url)
        return download?.state == Download.State.PENDING || download?.state == Download.State.RUNNING
    }

    @Synchronized
    fun isDownloadingTempFile(path: File?): Boolean {
        if (path == null) return false

        downloadMap?.values?.let { downloads ->
            for (download in downloads) {
                if (path.absolutePath == download.filePath + ".tmp" && isDownloading(download.url)) {
                    return true
                }
            }
        }

        return false
    }

    @Synchronized
    fun getDownload(url: String?): Download? {
        if (url == null) return null

        downloadMap?.values?.let { downloads ->
            for (download in downloads) {
                if (url == download.url) return download
            }
        }

        return null
    }

    @Synchronized
    fun cancelDownload(url: String) {
        if (Config.DEBUG) Log.i(TAG, "Download cancelled: $url")

        val download = getDownload(url)
        if (download != null && isDownloading(url)) {
            downloadClient?.cancel(download.id)
            downloadMap?.remove(download.id)
            stopJob()
        }
    }

    @Synchronized
    private fun updateDownloadProgress(downloadId: Int, bytesWritten: Long, totalBytes: Long) {
        downloadMap?.get(downloadId)?.also { download ->
            download.state = Download.State.RUNNING
            download.bytesWritten = bytesWritten
            download.totalBytes = totalBytes
        }

        var bytesWrittenOfAll: Long = 0
        var totalBytesOfAll: Long = 0

        downloadMap?.values?.let { downloads ->
            for (download in downloads) {
                bytesWrittenOfAll += download.bytesWritten
                totalBytesOfAll += download.totalBytes
            }
        }

        // Race condition NullPointerException can be thrown
        val notification = notificationUtil
            ?.getDownloadRunningNotification(runningDownloadTitles)
            ?.setProgress(
                100,
                (bytesWrittenOfAll / (totalBytesOfAll / 100.0)).toInt(),
                false
            )
            ?.build()
        notificationUtil?.notify(
            NotificationUtil.DOWNLOAD_RUNNING_NOTIFICATION_ID,
            notification
        )
    }

    @Synchronized
    private fun updateDownloadSuccess(downloadId: Int) {
        downloadMap?.get(downloadId)?.let { download ->
            download.state = Download.State.SUCCESSFUL
            if (download.showNotification) {
                notificationUtil?.showDownloadCompletedNotification(download)
            }

            App.instance.state.download.of(download.url).completed()
        }
    }

    @Synchronized
    private fun updateDownloadFailure(downloadId: Int) {
        downloadMap?.get(downloadId)?.also { download ->
            download.state = Download.State.FAILURE
        }
    }

    // Handler that receives messages from the thread
    private inner class ServiceHandler internal constructor(looper: Looper) : Handler(looper) {

        override fun handleMessage(message: Message) {
            val title = message.data.getString(ARG_TITLE)
            val url = message.data.getString(ARG_URL)
            val filePath = message.data.getString(ARG_FILE_PATH)
            val showNotification = message.data.getBoolean(ARG_SHOW_NOTIFICATION)

            var allowedNetworkTypes = DownloadRequest.NETWORK_WIFI or DownloadRequest.NETWORK_MOBILE
            val appPreferences = ApplicationPreferences()
            if (appPreferences.isDownloadNetworkLimitedOnMobile) {
                allowedNetworkTypes = DownloadRequest.NETWORK_WIFI
            }

            val request = DownloadRequest.Builder()
                .url(url)
                .retryTime(1)
                .progressInterval(1, TimeUnit.SECONDS)
                .allowedNetworkTypes(allowedNetworkTypes)
                .destinationFilePath(filePath)
                .downloadCallback(object : DownloadCallback() {
                    override fun onStart(downloadId: Int, totalBytes: Long) {
                        if (Config.DEBUG) Log.i(TAG, "Download started: $url")

                        updateDownloadProgress(downloadId, 0, totalBytes)
                    }

                    override fun onProgress(downloadId: Int, bytesWritten: Long, totalBytes: Long) {
                        updateDownloadProgress(downloadId, bytesWritten, totalBytes)
                    }

                    override fun onSuccess(downloadId: Int, filePath: String?) {
                        if (Config.DEBUG) Log.i(TAG, "Download finished: $filePath")

                        updateDownloadSuccess(downloadId)

                        stopJob()
                    }

                    override fun onFailure(downloadId: Int, statusCode: Int, errMsg: String?) {
                        Log.e(TAG, "Download failed: $errMsg")

                        updateDownloadFailure(downloadId)

                        stopJob()
                    }
                })
                .build()

            downloadClient?.let { downloadClient ->
                val download = Download()
                download.title = title
                download.url = url
                download.filePath = filePath
                download.showNotification = showNotification
                download.id = downloadClient.add(request)

                downloadMap?.putIfAbsent(download.id, download)
            }
        }

    }

}
